Adds writeOut function and restructure

7oi opened the pull request:
## Premise
I wanted to write junit reports to a file using gulp-lesshint so only logging to console was not optimal. So I looked at the behaviour of other reporters (lesshint-lint-xml-reporter, which borrowed from gulp-jshint-xml-file-reporter) and added that functionality.

## Changes
- Adds writeOut function, which writes out to file if filePath is given. Otherwise to console.
- comment and restructure module
- update version

Commit:
Adds writeOut function and restructure

- writeOut function writes out to file if filePath is given. Otherwise to console
- comment and restructure module
- update version

7oi commented:
@drinkataco Have you managed to check this out? I know it's a bit of a rewrite (I kind of forgot myself in it), but should function the same just with the added functionality of being able to write the report directly to file.

drinkataco commented:
Hi @7oi, thanks for your contribution.

The writeOut method is definitely needed, so big thanks for adding it.

Just a few issues however, it seems that the test is now failing `yarn test`, and the updates don't follow the style guide `yarn lint`.

I'll be sure to add these tests to the build process, as well as contributing information for the future soon

7oi commented:
Ah, so I see. I didn't look further into the Travis output, as it actually seems like it was passing. I also didn't realize there was a style guide defined (I'm quite new to node modules, tbh). I will fix these issues tomorrow and update the PR.

drinkataco commented:
@7oi I didn't put testing/linting in travis 🤦‍♂️, I'll be sure to fix the build

Commit:
Fix lint errors and clean up

Commit:
Fix failing tests and add test for writeout

Commit:
More tests for writeOut

Commit:
Run linting job in travis

Commit:
Adds mkdirp to dependencies

Commit:
Adds sinon to tests to grab exit code...

...as to not to override process.exit for mocha. We need the exit code to fail builds.

7oi commented:
@drinkataco I have updated the pull request. Here's what happened:
- fixed the tests that were breaking
- the code is now compliant with the linter
- added tests for the writeOut function
- added the linting step to .travis.yml
- added sinon to tests to grab exit code. The previously used method resulted in mocha always exiting with code 0, thus not failing the travis build on failed tests.

Commit:
Fix comments

drinkataco commented:
@7oi Sorry I haven't much time to check and merge this PR. I had a quick look tonight, and it seems to me that the tests aren't running correctly (and always exit as an ok) although I haven't been able to delve too much in this. I will prioritise in checking and merging this this week. thanks for your contribution

drinkataco commented on test/test.js:
```diff
@@ -76,28 +78,56 @@ describe('JUNIT Reporter', () => {
+    process.exit.restore();
+
+    // Wait 100 ms before checking if file exists, as writeOut is asynchronous
+    setTimeout(() => {
```
Shouldn't we use a promise here if asynchronous, instead of a timeout?

7oi replied:
Of course. I don't know why I didn't do that originally. I have updated the test.

Commit:
Use promise instead of setTimeout.
